### ISSUE TITLE
Certificate errors when using a proxy

### ISSUE THREAD

**Issue by jszwedko** [ansible/ansible#13024]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.9.4
  configured module search path = None
##### Ansible Configuration:

Nothing
##### Environment:

Debian 8.2
Python 2.7.9
##### Summary:

Very similar to #12549

Using a proxy for HTTPS endpoints causes Ansible to complain that the certificate doesn't match the hostname (I added some print statements to the generated python file to reveal that the hostname it tries to compare to is that of the proxy).

It works with Ansible 1.9.2, but not 1.9.3 or 1.9.4.
##### Steps To Reproduce:

```
https_proxy=http://10.41.16.10:3128 ansible -m get_url -a "url=https://www.google.com force=yes dest=/dev/null" -i localhost, all -c local -vv
```
##### Expected Results:

File is fetched without issue.
##### Actual Results:

```
<localhost> REMOTE_MODULE get_url url=https://www.google.com force=yes dest=/dev/null
localhost | FAILED >> {
    "failed": true,
    "msg": "SSL Certificate does not belong to www.google.com.  Make sure the url has a certificate that belongs to it or use validate_certs=False (insecure)"
}
```

**Comment by jszwedko** [comment]
cc/ @2rs2ts

**Comment by jszwedko** [comment]
Bisected down to this commit: a4691991ad83bccf6fb42bdb06a5da820e8a766e

**Comment by chrrrles** [comment]
Hey @abadger - do you have any insight on this?

**Comment by abadger** [comment]
@jszwedko is there any reason to believe this isn't the same as issue #12549 ?  (I just don't want to miss anything important in this report)

**Comment by jszwedko** [comment]
@abadger it is likely the same issue, but I just wanted to file against the current stable version of ansible as my understanding is that`2.0.0` codebase (which #12549 is filed against) has diverged quite a bit.

If you are confident that they are the same issue, I'm cool with closing this in lieu of the other and adding 1.9.3 and 1.9.4 as other affected versions.

**Comment by chrrrles** [comment]
I am unable to verify this on 2.0 or stable using trafficserver as a proxy:

```
[user@foo 12549]$ https_proxy=http://localhost:8080 ansible -m get_url -a "url=https://www.google.com force=yes dest=/
dev/null" -i localhost, all -c local -vv                                                                                             
<localhost> REMOTE_MODULE get_url url=https://www.google.com force=yes dest=/dev/null
localhost | success >> {
    "changed": true, 
    "checksum": "c91b3f5c500f3e7851f94ab239d78bd125cfbb0c", 
    "dest": "/dev/null", 
    "gid": 0, 
    "group": "root", 
    "md5sum": "d41d8cd98f00b204e9800998ecf8427e", 
    "mode": "0666", 
    "msg": "OK (unknown bytes)", 
    "owner": "root", 
    "secontext": "system_u:object_r:null_device_t:s0", 
    "sha256sum": "", 
    "size": 0, 
    "src": "/tmp/tmp4HsJiK", 
    "state": "file", 
    "uid": 0, 
    "url": "https://www.google.com"
}

[user@foo 12549]$ python --version
Python 2.7.9
[user@foo 12549]$ ansible --version
ansible 1.9.4 (stable-1.9 2d914d4b1e) last updated 2015/11/19 16:12:43 (GMT +800)
  lib/ansible/modules/core: (stable-1.9 4b65a4a8b5) last updated 2015/11/19 16:14:06 (GMT +800)
  lib/ansible/modules/extras: (stable-1.9 29c3e31a92) last updated 2015/11/19 16:14:16 (GMT +800)
  configured module search path = None
```

I am continuing to investigate, but is there any more information I may use to duplicate this bug?

**Comment by jszwedko** [comment]
@chrrrles Thanks for investigating! I'll get a version number for you tomorrow, but we were using Squid as our proxy (I believe it was the latest version, 3.5.11).

**Comment by jszwedko** [comment]
It was a little older:

```
% /usr/sbin/squid3 -v
Squid Cache: Version 3.4.8
 linux
configure options:  '--build=x86_64-linux-gnu' '--prefix=/usr' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--libexecdir=${prefix}/lib/squid3' '--srcdir=.' '--disable-maintainer-mode' '--disable-dependency-tracking' '--disable-silent-rules' '--datadir=/usr/share/squid3' '--sysconfdir=/etc/squid3' '--mandir=/usr/share/man' '--enable-inline' '--disable-arch-native' '--enable-async-io=8' '--enable-storeio=ufs,aufs,diskd,rock' '--enable-removal-policies=lru,heap' '--enable-delay-pools' '--enable-cache-digests' '--enable-icap-client' '--enable-follow-x-forwarded-for' '--enable-auth-basic=DB,fake,getpwnam,LDAP,MSNT,MSNT-multi-domain,NCSA,NIS,PAM,POP3,RADIUS,SASL,SMB' '--enable-auth-digest=file,LDAP' '--enable-auth-negotiate=kerberos,wrapper' '--enable-auth-ntlm=fake,smb_lm' '--enable-external-acl-helpers=file_userip,kerberos_ldap_group,LDAP_group,session,SQL_session,unix_group,wbinfo_group' '--enable-url-rewrite-helpers=fake' '--enable-eui' '--enable-esi' '--enable-icmp' '--enable-zph-qos' '--enable-ecap' '--disable-translation' '--with-swapdir=/var/spool/squid3' '--with-logdir=/var/log/squid3' '--with-pidfile=/var/run/squid3.pid' '--with-filedescriptors=65536' '--with-large-files' '--with-default-user=proxy' '--enable-build-info= linux' '--enable-linux-netfilter' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wall' 'LDFLAGS=-fPIE -pie -Wl,-z,relro -Wl,-z,now' 'CPPFLAGS=-D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security'
```

**Comment by abadger** [comment]
@jszwedko I've pushed a fix for #12549 into devel and stable-2.0 branches.  If you're able to test a checkout of that it would be great.  I'm not yet sure if we'll backport the fix to 1.9.x or leave it in 2.x only.
